applications/minotari_node/src/grpc/base_node_grpc_server.rs

```diff
@@ -788,27 +788,33 @@ impl tari_rpc::base_node_server::BaseNode for BaseNodeGrpcServer {
         let mut coinbases: Vec<tari_rpc::NewBlockCoinbase> = request.coinbases;
 
         // let validate the coinbase amounts;
-        let reward = self
-            .consensus_rules
-            .calculate_coinbase_and_fees(new_template.header.height, new_template.body.kernels())
-            .map_err(|_| {
-                obscure_error_if_true(
-                    report_error_flag,
-                    Status::internal("Could not calculate the amount of fees in the block".to_string()),
-                )
-            })?
-            .as_u64();
-        let mut total_shares = 0u64;
+        let reward = u128::from(
+            self.consensus_rules
+                .calculate_coinbase_and_fees(new_template.header.height, new_template.body.kernels())
+                .map_err(|_| {
+                    obscure_error_if_true(
+                        report_error_flag,
+                        Status::internal("Could not calculate the amount of fees in the block".to_string()),
+                    )
+                })?
+                .as_u64(),
+        );
+        let mut total_shares = 0u128;
         for coinbase in &coinbases {
-            total_shares += coinbase.value;
+            total_shares += u128::from(coinbase.value);
         }
-        let mut remainder = reward - ((reward / total_shares) * total_shares);
+        let mut cur_share_sum = 0u128;
+        let mut prev_coinbase_value = 0u128;
         for coinbase in &mut coinbases {
-            coinbase.value *= reward / total_shares;
-            if remainder > 0 {
-                coinbase.value += 1;
-                remainder -= 1;
-            }
+            cur_share_sum += u128::from(coinbase.value);
+            coinbase.value =
+                u64::try_from((cur_share_sum * reward) / total_shares - prev_coinbase_value).map_err(|_| {
```

Comment:
Any issue with possible overflow in the multiplication, division by zero, or underflow in the subtraction?

Reply:
We could test for that, and subdue future comments.

Reply:
over flow should not be possible as its all u64, but I made them checked and saturating anyway